feat: Secret-store implementation

srebhan opened the pull request:
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #3124
replaces: #9775 

This PR implements a secret-store for storing sensitive data such as usernames, passwords, tokens, etc. Those secrets can then be referenced in the Telegraf config with a `@{store_id:secret key}` pattern. This effectively allows to remove secrets from config file(s) and store them in a "safe" place.

The secret-store can be configured using `[[secretstores.xyz]]` sections where `xyz` references the secret-store plugin to use. Currently, implementations to use OS native stores ([kernel keyrings](https://man7.org/linux/man-pages/man7/keyrings.7.html) for linux, [Windows credential manager](https://support.microsoft.com/en-us/windows/accessing-credential-manager-1b5c916a-6a16-889f-8581-fc16e8165ac0) and [keychain](https://developer.apple.com/documentation/security/keychain_services) for MacOS) and Javascript Object Signing and Encryption files are available. Further implementation may be added later.

The stored secrets can be easily referenced via an `@{<secretstore id>:<secret identifier>}` pattern, with `<secretstore id>` being the *unique identified* you give to the secret-store and `secret identifier` being the *unique key* for the secret in that store. Multiple secret-stores can be defined and referenced by the `id` specified in the secret-store config.
To enable secret references in a config option you need to change the type to `config.Secret`, which will also make static secrets stored encrypted in memory. This has be done in this PR for `inputs.http` and `inputs.sql` as examples.

Secrets can be either managed using the tools native to the backend (e.g. using [keyctl](https://linux.die.net/man/1/keyctl)) or using Telegraf. Commands for getting/setting and listing secrets have been added (e.g. `./telegraf secret set <secretstore name> sqlthing "incredible"`). When using Telegraf, the config is parsed and _must_ contain an uncommented `[[secretstores.xyz]]` section.

**Limitations**

- There is no access control, i.e. if a plugin references a secret it will _get_ that secret no matter if it was intended by the user as secrets are simply injected into the structure. We might work out an access-control regime here, but it is not yet in place (suggestions welcome).
- Currently, all `@{\w+:[\w ]+}` patterns are replaced (in the tagged fields). There is currently no way to escape that pattern.
- tested with Linux _kernel user-session keyring_ and _file_ keyring _ONLY_

Suggestions, comments and critique is welcome!

jdstrand commented:
I skimmed the PR and may have missed a detail, but how are secrets meant to survive reboots? I saw the section on JOSE, but then the kernel keyring seemed separate from that. Perhaps put another way: can you provide user journeys on how this is meant to be used along with lifecycle (create, modify, use, revoke, reboot, etc) discussion (all with pointers into implementation details)? Perhaps you have a spec that outlines all of this already. Thanks!

reimda reviewed (commented):
Some initial questions

srebhan commented:
> I skimmed the PR and may have missed a detail, but how are secrets meant to survive reboots? I saw the section on JOSE, but then the kernel keyring seemed separate from that.

The secrets in the linux secret-store will not survive a reboot. You need to re-inject them after boot-up. However, this is a limitation of the kernel keyring and out-of-scope for Telegraf.

> Perhaps put another way: can you provide user journeys on how this is meant to be used along with lifecycle (create, modify, use, revoke, reboot, etc) discussion (all with pointers into implementation details)?

I'm not sure what you want me to do here, but I'm a bit hesitant to document how to setup and maintain a kernel keyring as it a general Linux thing, tools might change, etc...

srebhan commented on config/config.go:
```diff
@@ -77,7 +78,9 @@ type Config struct {
 
-	Agent       *AgentConfig
+	Agent        *AgentConfig
+	SecretStores map[string]telegraf.SecretStore
```
Hmmm, I don't get your point yet. The `SecretStores` map holds the mapping of the _unique_ store ID to the instance. So you can use and mix as many secret-stores as you like but need to make sure it has a _unique_ ID. What am I missing?

srebhan commented on cmd/telegraf/telegraf.go:
```diff
@@ -482,8 +480,113 @@ func main() {
+				log.Fatal("E! No secretstore configured in config!")
+			}
+
+			switch args[1] {
```
Please have a look at my attempt to add a help text. Feedback is welcome!

jdstrand commented:
> > I skimmed the PR and may have missed a detail, but how are secrets meant to survive reboots? I saw the section on JOSE, but then the kernel keyring seemed separate from that.
> 
> The secrets in the linux secret-store will not survive a reboot. You need to re-inject them after boot-up. However, this is a limitation of the kernel keyring and out-of-scope for Telegraf.
> 
> > Perhaps put another way: can you provide user journeys on how this is meant to be used along with lifecycle (create, modify, use, revoke, reboot, etc) discussion (all with pointers into implementation details)?
> 
> I'm not sure what you want me to do here, but I'm a bit hesitant to document how to setup and maintain a kernel keyring as it a general Linux thing, tools might change, etc...

I wasn't looking for you to add to the docs how to manage all this stuff. I was trying to understand the feature in the first place. I'm aware that the kernel keyring doesn't survive reboots, but the feature seems to imply removing secrets out of the config files and so I wondered who wants this feature and why. I'm unclear on what the expected experience is intended to be. Your answer implies that it is up to the user to put things where they need to be and to tell telegraf where they are, but I didn't glean that from this PR, which is why I asked for user journeys.

jdstrand commented:
> I'm aware that the kernel keyring doesn't survive reboots, but the feature seems to imply removing secrets out of the config files and so I wondered who wants this feature and why. I'm unclear on what the expected experience is intended to be. Your answer implies that it is up to the user to put things where they need to be and to tell telegraf where they are, but I didn't glean that from this PR, which is why I asked for user journeys.

Replying to myself: I spoke with someone out of band and indeed, this is the requested feature. Specifically, the request is not for a complete solution in telegraf itself for secure storage, but instead to have a building block for secure secrets storage that the user can use as part of a larger solution where telegraf simply fetches secrets from a configured-outside-of-telegraf secure location (like the kernel keyring, etc).

srebhan commented:
Sorry for the late answer. Indeed, the idea is to be able to connect to other secret stores like Vault, AWS, etc, the current PR provides the infrastructure and two simple bare-minimum plugins for an MVP.
The other target point is to move secrets out of the telegraf config to be able to check it into code-versioning systems as usually policies of some users might not allow those files to contain secrets/credentials.

Hope this makes more sense now @jdstrand!?

jdstrand reviewed (changes requested):
I have not completed my review but came across a few things that I wanted to highlight right away. See inline (in addition to the documentation updates, note that I wasn't able to successfully use `@{id:secret_key}` in my telegraf config TOML.

srebhan commented on plugins/secretstores/jose/README.md:
```diff
@@ -0,0 +1,27 @@
+# File based Javascript Object Signing and Encryption based secret-store
+[[secretstores.jose]]
+  ## Unique Identifier for the secret-store used to reference the secrets later via @{id:secret_key} (mandatory)
+  id = "secretstore"
```
Good idea, will do this and try to make the comment more clear.

srebhan commented on docs/COMMANDS_AND_FLAGS.md:
```diff
@@ -66,3 +67,39 @@ telegraf [flags]
+
+To access the secret *values* you can use the `get` command
+
+`telegraf secret get someid`
```
I'll have to look into this! Thanks for looking into it. Can you maybe also show the configuration of `"secretstore"` in `telegraf.conf`?

srebhan replied:
I see the issue now. There can be only one password per `jose` secret-store. So having two secrets with different passwords in _the same_ store does not work. That's why it is asking for a password for `/dev/shm/secrets` without appending the secret name, because it really wants the per-store password.

I'm not sure on how to document this, because this is a "feature" of the underlying library and might or might not change in the future... Any idea or suggestion?

srebhan replied:
I see the issue. The current design is that there should be one password/secret to unlock a secret store as we have no way to specify the exact secret-key for a password. That's why there is only _one_ `password` option per secret-store.

This being said, with `jose` we have no way of testing if you are using the same password. It gets even more weird as
you can use
```toml
[[secretstores.jose]]
  id = "test"
  path = "/dev/shm/secrets"

[[secretstores.jose]]
  id = "test1"
  path = "/dev/shm/secrets"
  password = "pass1"

[[secretstores.jose]]
  id = "test2"
  path = "/dev/shm/secrets"
  password = "pass2"
```
and later reference
```toml
  username = "@{test1:name1}"
  password = "@{test2:name2}"
```

with your example and it will work.

The issue is that I have no way of finding out if _all_ secrets in the subdirectory have the same password to unlock...
Any suggestion?

jdstrand replied:
One way to handle this would be to avoid it in the first place. Eg:
* when there are no secrets in the `path`, `telegraf secret set ...` simply adds it
* when there are 1 or more secrets in the `path`, `telegraf secret set ...` first tries to unlock all of the existing, and if it can't, it errors with a useful error

srebhan replied:
This is now completely switched to urfave CLI having single-word subcommands (`list-secretstores`, `list-secrets`, `get-secret` and `set-secret`).

srebhan commented on config/secret.go:
```diff
@@ -0,0 +1,196 @@
+var unlinkedSecrets = make([]*Secret, 0)
+
+// secretPattern is a regex to extract references to secrets stored in a secret-store.
+var secretPattern = regexp.MustCompile(`@\{(\w+:[\w ]+)\}`)
```
I can remove the space... Would that be better?

jdstrand replied:
Being a UNIX person, spaces look weird to me for keys in key/value situations. It isn't a security concern; I just thought it was inconsistent that the store id didn't allow for a space but the secret name did. Personally without knowing the requirements, I'd leave out the space for both, but take that for what it's worth (personal preference).

srebhan replied:
Removed space.

jdstrand commented on docs/COMMANDS_AND_FLAGS.md:
```diff
@@ -66,3 +67,39 @@ telegraf [flags]
+
+to add or overwrite a secret names `a_secret_key` with the value `the_new_value` in the
+secret-store with ID `someid`. To later reference that secret in your configuration you
+can use `@{someid:a_secret_key}`.
```
Ah! That explains it. :)

srebhan replied:
I will close this as resolved for now as it is out-of-scope for this PR to convert all plugins.

srebhan commented on plugins/secretstores/jose/README.md:
```diff
@@ -0,0 +1,27 @@
+To access the secrets, a password is required. This password can be specified using the `password`
+parameter containing a string, an environment variable or as a reference to a secret in another
+secret store. If `password` is not specified in the config, you will be prompted for the password
+at startup.
```
Linked to the `telegraf` options to manage secrets in each plugin README.

jdstrand reviewed (changes requested):
I finished my review of this PR. Please note, I am not a cryptographer and I have not done a full source code audit of the added dependencies, but I have spot-checked a few things.

The `github.com/99designs/keyring` project is used by telegraf's secret store as a standard interface for different credential storage services. For this PR, the `keyring` package is used to implement backends for file-backed, JOSE credential storage as well as native OS credential storage (`keyctl` on Linux, `keychain` for OSX and `wincred` for Windows). These backends are implemented as secret store telegraf plugins, and these plugins implement the new `SecretStore` interface and these implementations use `keyring` internally. This project appears to be healthy and used by a few folks, with 361 stars, 56 watchers and 98 forks. It is used be the popular part of `github.com/99designs/aws-vault` project (429k downloads, 6,500 stars, 123 watchers, 637 forks). `github.com/99designs/keyring` has < 2500 lines of code with reasonable documentation.

The `github.com/awnumar/memguard` project is used in an effort to securely handle secrets obtained from the secret store. The project appears to be healthy and fairly popular with 2,200 stars, 50 watchers and 102 forks. It has < 4000 lines of code. It has good documentation and between that and spot checking the code, the author has written this with security in mind. https://github.com/awnumar/memguard#features in particular discusses a number of features of the project.

At a high level, `telegraf` config initializes any configured secret stores and stores and retrieves secrets from the aforementioned `keyring` backends into protected memory via `memguard`. This PR initially sets up the `http` and `sql` input plugins to optionally use a configured secret store.

Please see inline for a discussion of how `config/secret.go:read()` improperly copies a protected secret into standard memory. For `memguard` to have its intended effectiveness, this should be addressed.

Please see inline in the `go.sum` diff for a discussion of the added dependencies. Nothing of great concern, but worth understanding.

Please see inline for a discussion of desired missing test cases. Eg, `secret.go` lacks test coverage. People understandably have different opinions on 100% coverage and I'm not advocating for that per se (I wouldn't object to it though!), however, I think with something as sensitive as secrets, it might make sense to expand tests to various error conditions/etc where possible since sometimes weird things can happen in these edge cases.

I did manual functional testing for the JOSE and OS/Linux secret-store usage and gave feedback in my initial, abbreviated review of this PR.

## JOSE

The JOSE backend works well enough with `telegraf secret set secretstore-jose myname mysecret` creating a file-backed secret store at the configured directory `path` with file at `/path/for/config/myname`. The contents of the file are a base64 encoded JOSE file and encrypted to the password defined in the telegraf configuration (or via telegraf prompt) using `PBES2_HS256_A128KW / A256GCM`. Inspecting file directly, we are unable to see the secret:
```
$ cat /dev/shm/secrets/myname
eyJhbGciOiJQQkVTMi1IUzI1NitBMTI4S1ciLCJjcmVhdGVkIjoiMjAyMi0wOC0wOSAxOTozMDoxNC4yMTY2NDU0NjIgKzAwMDAgVVRDIG09KzAuMTAzNDgwNDMzIiwiZW5jIjoiQTI1NkdDTSIsInAyYyI6ODE5MiwicDJzIjoibGw4M0M4VzJfRWFCclZ4eCJ9.Fkwwup5DCL6HN98dLqyhJpNYqlNtwfA6dPCRqhpPUzBBPJJ892qCWA.YTRpBuU0bk6PMG4K.nXw0W1dBRnmjq-YAA68XJ4SICqTO7NTJXvjMmSVR9_XsMwLJfNQ9qSmqmgMpdjXblVIV4nY6f0EbkNQodzCDrd-uBqLewNs7XeteeCbyevIz-A6QQxpFy-4uU0Ai4rlm_WbcOnSJZrkBkAzjtOVIpWLwvtx-k5ZzFbuIkz5VIdgUOFCI.4BuDxZk0c_7n9fhcliTKWg

$ for s in $(cat /dev/shm/secrets/myname | tr '.' '\n') ; do echo "# $s" ; echo "$s" | base64 -d ; echo ; echo ; done
# eyJhbGciOiJQQkVTMi1IUzI1NitBMTI4S1ciLCJjcmVhdGVkIjoiMjAyMi0wOC0wOSAxOTozMDoxNC4yMTY2NDU0NjIgKzAwMDAgVVRDIG09KzAuMTAzNDgwNDMzIiwiZW5jIjoiQTI1NkdDTSIsInAyYyI6ODE5MiwicDJzIjoibGw4M0M4VzJfRWFCclZ4eCJ9
{"alg":"PBES2-HS256+A128KW","created":"2022-08-09 19:30:14.216645462 +0000 UTC m=+0.103480433","enc":"A256GCM","p2c":8192,"p2s":"ll83C8W2_EaBrVxx"}

# Fkwwup5DCL6HN98dLqyhJpNYqlNtwfA6dPCRqhpPUzBBPJJ892qCWA
L0����7�.��&�X�Sm��:t��OS0A<�|�j�Xbase64: invalid input

# YTRpBuU0bk6PMG4K
a4i�4nN�0n

# nXw0W1dBRnmjq-YAA68XJ4SICqTO7NTJXvjMmSVR9_XsMwLJfNQ9qSmqmgMpdjXblVIV4nY6f0EbkNQodzCDrd-uBqLewNs7XeteeCbyevIz-A6QQxpFy-4uU0Ai4rlm_WbcOnSJZrkBkAzjtOVIpWLwvtx-k5ZzFbuIkz5VIdgUOFCI
�|4[WAFy�base64: invalid input

# 4BuDxZk0c_7n9fhcliTKWg
�ř4base64: invalid input
```

Obtaining the keys with `telegraf secret get secretstore-jose` works as expected (requires password in the telegraf config to match the password that was used to encrypt) and returns `mysecret`.

The JOSE backend supports multiple secrets (though see my prior review for edge cases).

## OS

I only manually tested the Linux `keyctl` backend. The OS backend works well enough with `telegraf secret set secretstore-os myname mysecret` creating entries in the kernel 'user keyring':

```
$ telegraf secret set secretstore-os myname mysecret2
2022-08-09T20:27:03Z I! Using config file: ../tg-secret.conf

$ keyctl show @u
Keyring
 361220595 --alswrv   1000 65534  keyring: _uid.1000
 814727799 --alswrv   1000  1000   \_ keyring: telegraf
 195945450 --alswrv   1000  1000       \_ user: myname

$ keyctl request user myname
195945450

$ keyctl print 195945450
mysecret2
```

Obtaining the keys with `telegraf secret get secretstore-os` works as expected and returns `mysecret2` (note, no password is required, see below).

The OS/Linux backend supports multiple secrets.

Management of keys is done via syscalls and therefore mediated by various kernel mechanisms. Keys have an owning user id, owning group and security label. Each key also has a set of permissions which is beyond normal UNIX file permissions. Keys require unswappable kernel memory and are quota controlled. Keys may have an expiration.

Keys may be of different types: keyring, user, etc.
* "user" is a general purpose key type and always kept in kernel memory. It is accessible for read and write from userspace.
* "keyring" is a special key which stores links to other keys, similar to a directory holding links to files. Keys need to be anchored to not be garbage collected and keyrings are used to anchor keys. Types of anchor keyrings:
  * process keyrings: process credentials reference keyrings with specific semantics. 3 keyrings with different inheritance/shared: session-keyring(7) (inherited/shared by all child processes), process-keyring(7) (shared by all threads in a process) and thread-keyring(7) (specific to thread)
  * user keyrings: each UID has two keyrings: user-keyring(7) and user-session-keyring(7) and they exist for as long as the kernel knows about the UID record
  * persistent keyrings: a persistent-keyring(7) is available to each UID known to the system and may persist beyond the life of the UID record, but with an expiration
  * special keyrings: used by the kernel for special purposes, et, a system keyring for holding encryption keys for module signature verification. Typically not alterable by user space
  * unimplemented 'group keyring'

Keys have 'possession' rules and permissions. (see man keyrings)

With the telegraf implementation, it anchors a user key into a user keyring that is specific to the user id, so telegraf running as the telegraf user storing keys in the user keyring would have the key protected from other users on the system. It has view, read, write, search, link and setattr permissions for possessor and owner:
```
$ keyctl show @u
Keyring
 361220595 --alswrv   1000 65534  keyring: _uid.1000
 814727799 --alswrv   1000  1000   \_ keyring: telegraf
 195945450 --alswrv   1000  1000       \_ user: myname

# the 'alswrv' corresponds to setattr, link, search, write, read, view. There
# are 4 groups of these 6 characters that correspond to: possessor, owner, group
# and other.
$ keyctl describe 195945450
195945450: alswrvalswrv------------  1000  1000 user: myname
```

This setup is similar to the DAC protections given via `/proc/<pid>/environ` (which has 0400 permissions and mediated by the `PTRACE_MODE_READ_FSCREDS` capability).

Being in the kernel keyring, the kernel has additional access controls:
* there are LSM hooks for keyctl
* changing the ownership (uid and gid) of a key (`man 2 keyctl`, `KEYCTL_CHOWN`) requires `CAP_SYS_ADMIN`
* changing the permissions of the key (`man 2 keyctl`, `KEYCTL_SETPERM`) requires `CAP_SYS_ADMIN` for keys that the user does not own.

`/proc/keys` is 0444 and shows all keyring names that the user can access (based on internal kernel access controls). `/proc/key-users` is 0444 and show statistics for different keyring users.

In general, `telegraf` when configured to use OS on Linux is not prescribing a methodology to securely put keys into the kernel keyring for the telegraf user as part of a larger solution. Instead, it provides a means for `telegraf` to access them (and tooling to insert). This portion of the implementation alone is not significantly different from putting a variable in the environment for telegraf to use. What makes this PR interesting from a security perspective is the addition of memguard for these secrets (an improvement over today) and by virtue of using the kernel keyring, it allows users to populate in the manner they choose (such as trusted keys via TPM support, etc). It might be interesting to consider 'encrypted keys' in future work depending on requirements.

References:
* https://www.kernel.org/doc/Documentation/security/keys.txt
* https://www.kernel.org/doc/Documentation/security/keys-trusted-encrypted.txt
* man proc
* man capabilities
* man keyrings
* man 2 keyctl
* man 3 keyctl

# Summary
On the whole, this looks good overall. The choice of software seems reasonable and adding JOSE along with OS/Linux, OS/OSX and OS/Windows in the initial implementation of this PR seems useful to `telegraf` users.

I'd like to see more test coverage and consideration for secret handling after leaving memguard. See inline and my previous review for other comments.

srebhan commented on config/secret.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+// UnmarshalTOML creates a secret from a toml value
+func (s *Secret) UnmarshalTOML(b []byte) error {
```
This is a interface requirement by the TOML implementation to be able to unmarshal the configuration to the actual structure. There is nothing we can do about the naming if we want to be able to read secrets from the configuration.

jdstrand replied:
That's fine. A comment would help. `init()` doesn't imply `unmarshall` to me, but it is calling `unquote()` which is essentially doing that. Eg, maybe:
```
...
// init() unmarshalls and puts the secret into a protected memory
err := s.init(b)
...
```

srebhan commented on config/secret.go:
```diff
@@ -0,0 +1,196 @@
+)
+
+// unlinkedSecrets contains the list of secrets that must be linked by the config.
+var unlinkedSecrets = make([]*Secret, 0)
```
These are the secrets that reference a secret store (in contrast to those using an environment variable or being a string). At the point of configuration loading we cannot resolve the reference as the configuration might span multiple files and the secret-store configuration might come after the reference during loading. Therefore, we first need to store the references and then, after loading is completed, link the secrets to the secret stores...

Do you have a better naming (me being bad at names)? Anyway, I should put the comment above next to this declaration...

jdstrand replied:
Perhaps expanding on the comment either here or in another appropriate location to include a bit of what you just described would be sufficient for future maintainers.

srebhan commented on go.sum:
```diff
@@ -84,6 +84,9 @@ collectd.org v0.5.0/go.mod h1:A/8DzQBkF6abtvrT2j/AU/4tiBgJWYyh0y/oB/4MlWE=
 dmitri.shuralyov.com/gpu/mtl v0.0.0-20190408044501-666a987793e9/go.mod h1:H6x//7gZCb22OMCxBHrMx7a5I7Hp++hsVxbQ4BYO7hU=
 gioui.org v0.0.0-20210308172011-57750fc8a0a6/go.mod h1:RSH6KIUZ0p2xy5zHDxgAM4zumjgTw83q2ge/PI+yyw8=
+github.com/99designs/keyring v1.1.7-0.20211011120927-33891b18a1fa h1:3LyIWbLY03gktapLPoOMdtISxb6FyKMpWNUrjjEYT8Y=
+github.com/99designs/keyring v1.1.7-0.20211011120927-33891b18a1fa/go.mod h1:zTekYv1iRUODu1CH/T2GVzta6sJceVcTI7CWJ1FwNX0=
```
Will look into those. However, this will require changes in the dependency projects that we do not control.

jdstrand replied:
IME, you don't need to drop these since they aren't very complicated. I just wanted to call out that if there were bugs in this code that affected telegraf, it is likely that we would have to patch them ourselves.

srebhan commented on config/secret.go:
```diff
@@ -0,0 +1,196 @@
+	var buf bytes.Buffer
+	_, err = io.Copy(&buf, lockbuf.Reader())
+
+	return buf.Bytes(), err
```
The reasoning was the following. The `read` function shown above tries to guarantee the 

>  After use, the LockedBuffer should be destroyed.

part (see `defer lockbuf.Destroy()`). It is used in two places, the `Link` part and the `Get` part. In both places, we need the unencrypted secret to either link it to the secret stores (or replace static parts) or to return it to the plugin for further use (e.g. to add it to a query or authentication call). Unfortunately, we cannot return the `lockbuf` Reader, as it is not valid after destruction. Returning it to the plugin or `Link` caller is not an option, as we cannot guarantee the `lockbuf` is actually destroyed, running into instance limitations.

This leaves us with copying the `read` code to the two calling functions (`Link` and `Get`). For both functions, this has the gain to safe a placement on the stack, but the `lockbuf` destruction might be delayed and we have code duplication. Anyway, if you prefer to have it done this way, I can change the code.

In all cases, we, at some point, need to make the unencrypted secret accessible to the plugin for use. As most of the secrets are embedded into authentication calls requiring the plain-text password as string, this is unavoidable and will always lead to places where the secret is stored in unprotected memory. The memguard protection just pushes this part a bit down the pipe at least _enabling _ plugins to handle secrets safely.

jdstrand replied:
> The reasoning was the following. The `read` function shown above tries to guarantee the
> 
> > After use, the LockedBuffer should be destroyed.
>
> part (see `defer lockbuf.Destroy()`). It is used in two places, the `Link` part and the `Get` part. In both places, we need the unencrypted secret to either link it to the secret stores (or replace static parts) or to return it to the plugin for further use (e.g. to add it to a query or authentication call). Unfortunately, we cannot return the `lockbuf` Reader, as it is not valid after destruction. Returning it to the plugin or `Link` caller is not an option, as we cannot guarantee the `lockbuf` is actually destroyed, running into instance limitations.

Yes, I saw this which was why I mentioned that perhaps there was a pattern you could employ that would return an implementation of an interface that would have a method that would both read the secret out and then destroy the `LockedBuffer`, which would satisfy this recommendation and make that method sorta 'one time use'. Note, I was spit-balling here.

> This leaves us with copying the `read` code to the two calling functions (`Link` and `Get`). For both functions, this has the gain to safe a placement on the stack, but the `lockbuf` destruction might be delayed and we have code duplication. Anyway, if you prefer to have it done this way, I can change the code.
> 
> In all cases, we, at some point, need to make the unencrypted secret accessible to the plugin for use. As most of the secrets are embedded into authentication calls requiring the plain-text password as string, this is unavoidable and will always lead to places where the secret is stored in unprotected memory. The memguard protection just pushes this part a bit down the pipe at least _enabling _ plugins to handle secrets safely.

Reading the code, I don't think this is on the stack since a [Buffer is a variable-sized buffer of bytes with Read and Write methods](https://pkg.go.dev/bytes#Buffer). Even if it were on the stack, while it would be safe from the GC, it could still be [dumpable](https://man7.org/linux/man-pages/man5/core.5.html) and [swapped](https://www.kernel.org/doc/html/latest/admin-guide/mm/concepts.html#anonymous-memory) (at least on Linux).

It's understood that the unencrypted secret must be made available to the plugin to use. What I was trying to think through was the lifetime of the unencrypted secret. `memguard`'s goals are to "reduce the likelihood of sensitive data being exposed when in memory" with the idea that you leave it in `memguard`'s enclave as often as is possible and only pull it out when absolutely needed.

The way that I thought about this wrt `telegraf` was that once the secret was fetched from secure storage and placed under `memguard`, the plugins would ask `memguard` to decrypt the secret *every time they needed it and then immediately discard it* which could have a meaningful improvement of security *iff the plugins handled the decrypted secret safely* (which might include a [guarded allocation](https://github.com/awnumar/memguard/blob/master/examples/stream/stream.go#L42) but perhaps simply on the stack (to avoid GC; perhaps overwriting with 0s as part of a `defer` could be interesting as well...)). This is a meaningful improvement since the unencrypted secret is rarely in memory (in terms of CPU time) and therefore far less likely to be leaked unencrypted to disk. It is not perfect since there might be brief moments where it is exposed and of course if a debugger is attached (though, on Linux [`PR_SET_DUMPABLE`](https://man7.org/linux/man-pages/man2/prctl.2.html) is an option to prevent `ptrace`).

Looking at the code:
* `plugins/inputs/http/http.go:gatherURL()` fetches the username and password from the secure enclave when the URL is constructed, and it looks like [it is constructed each time `telegraf` runs the plugin](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/http/http.go#L71) (ie, the URL (and therefore the username and password) doesn't seem to be cached; please correct me. This is good!). However, `plugins/inputs/http/http.go:gatherURL()` holds onto the memory for the rest of the function instead of wiping it, etc after `h.client.Do(request)`. This is a missed opportunity, not to mention the allocation is on the heap (see above)
* `plugins/inputs/sql/sql.go:Init()` fetches the DSN from the secure enclave and doesn't cache it (good!) and then holds onto it for the remainder of the function. This is a missed opportunity, not to mention the allocation is on the heap (see above)
* `plugins/inputs/sql/sql.go:Start()` fetches the DSN from the secure enclave and then holds onto it via `s.db, err = dbsql.Open(s.driverName, dsn)` until `plugins/inputs/sql/sql.go:Stop()` is called (I didn't chase down if `s.db.Close()` (called from `Stop()`) let go of it; presumably it would). It appears that `Start()` and `Stop()` are called around `s.startGathering()` (based on `plugins/common/shim/input.go`); I didn't confirm this further. The DSN is in memory for quite some time with this plugin. If I'm reading correctly, for the entire time that all input plugins are running (please correct me; this was something I inferred)
* `config/secret.go:Link()` fetches the secret and holds onto it for long enough to check to see if the secret currently stored in `memguard` is the same as the one fetched from secret storage (this is good). The allocation is on the heap (see above) and therefore could be improved

The `http` plugin, `sql:Init()` and `Link()` are quite close to what I think is desired in that they are fetching when they need it, though I think there are improvements on the timing of when they 'let go' of the unencrypted secret and more importantly, that (I believe) the allocation is on the heap and will linger in the GC. Ensuring it is on the stack, scribbling over it with 0s/non-blocking random data immediately after use or (gasp) using CGO with `calloc` (then scribble 0s/non-blocking random data and `free` after use) might be options.

The architecture for `Start()` input plugins that need to start, gather all the inputs, then `Stop()` all the input plugins that need to stop is a bit problematic and is at odds with how we'd like to use `memguard`. I'm not sure what to suggest here (`Start()`ing, gathering and `Stop()`ing all the secret storage-using input plugins *one at a time* and then doing all the non-secret storage-using input plugins like today could be a way forward; there are likely other choices). If nothing is done with this plugin, it should probably at a minimum be called out in documentation as a limitation of this input plugin's use of secret storage.

srebhan replied:
First of all I moved the `read()` code into the two places where it was called and reduced the number of copies flying around. I cannot think of a way to reduce this further in those functions.
Regarding an interface function, I would rather not want that. While it might slightly extend the range of the encrypted secret it has multiple drawbacks making this very fragile as we give-up the control over the lockbuf

1. Only a limited number of `lockbuf` objects can exist at the same time (see [documentation](https://pkg.go.dev/github.com/awnumar/memguard#LockedBuffer)), so if the receiver holds a reference for too long we might run out of buffers.
2. As we do not have control over the potential implementation, we cannot guarantee that `lockbuf.Destroy()` is actually called, e.g. if the secret is never read.
3. A plugin can still copy the content of the lockbuf, reducing the overall gain at the cost of a higher complexity.

Regarding the plugins, I really appreciate your analysis and we should definitively look into those. However, I think the current PR cannot and should not try to solve all problems at once. Usually the plugins will call external function which usually do not make any attempt to secure credentials (database drivers, API implementations, etc). So let's push this down to the plugin level for now and *allow* for plugins which handle stuff securely, which is already a leap to the current state.

srebhan commented on plugins/secretstores/os/os.go:
```diff
@@ -0,0 +1,95 @@
+//go:build darwin || linux || windows
+// +build darwin linux windows
```
That's because it is impossible to mock those without _really_ interacting with the system. Not sure if we want this...

jdstrand replied:
I was more thinking that you could create an `os_test.go` that mocked return values for `createKeyringConfig()`, etc so you could drive unit tests in `os.go`. It is true that these files are rather straightforward, but the lack of any unit tests bothered me. If the `telegraf` team feels that the effort outweighs the benefit, I won't block on this. You could also add functional tests somewhere (AIUI, `telegraf` started using LXD for things; you could add some OS/Linux (keyctl) tests there...).

srebhan replied:
I added some tests that do not require modifying the system's secrets, so mostly error paths. Adding an integration test in a container is a good idea. Will follow-up on this.

jdstrand commented on config/types.go:
```diff
@@ -90,3 +91,19 @@ func (s *Size) UnmarshalTOML(b []byte) error {
+	if bytes.HasPrefix(b, []byte("\"")) && bytes.HasSuffix(b, []byte("\"")) {
+		return b[1 : len(b)-1]
+	}
+	return b
```
Yeah, if the TOML parser is going to fail earlier than here due to the various weird quoting examples I gave, then perhaps rename it and do something like:
```
// unquote() assumes valid TOML that guarantees we either have single or triple quotes
func unquoteTOML(b []byte) []byte {
```

srebhan replied:
Switched to `strconv.Unquote()`.

srebhan commented on config/secret.go:
```diff
@@ -0,0 +1,196 @@
+// UnmarshalTOML creates a secret from a toml value
+func (s *Secret) UnmarshalTOML(b []byte) error {
+	if err := s.init(b); err != nil {
+		return err
```
This cannot be tested as the return code is always `nil`. :-) Removed the return value.

jdstrand commented on plugins/secretstores/os/README.md:
```diff
@@ -0,0 +1,113 @@
+`opening keyring failed: Specified keyring backend not available` error!
+
+You can enable access to the kernel keyring, but as the keyring is __not__
+namespaced, you should be aware of the security implication!
```
You might want to highlight what one or two security implications are at a high level.

jdstrand commented on plugins/secretstores/os/sample_darwin.conf:
```diff
@@ -0,0 +1,17 @@
+  # password = ""
+
+  ## Allow dynamic secrets that are updated during runtime of telegraf
+  # dynamic = false
```
Are these example files expected to have newlines?

srebhan replied:
Yes. Fixed.

jdstrand reviewed (changes requested):
Thanks for the updates to the PR! I looked at the changes for addressing my feedback and things are looking really good. I did a cursory look over the whole PR and nothing jumped out, but am relying on the telegraf team to review those parts.

I also looked at the revamped CLI since my initial review and it works nicely.

I couldn't remember if I verified the file permissions with JOSE, so I did this time:
```
$ ls -l /dev/shm
total 0

$ telegraf set-secret secretstore myname mypass
2022/10/14 15:45:31 I! Using config file: ./tg-secret.conf
Enter passphrase to unlock "/dev/shm/secrets":

$ ls -lR /dev/shm
/dev/shm:
total 0
drwx------ 2 ubuntu ubuntu 60 Oct 14 15:45 secrets

/dev/shm/secrets:
total 4
-rw------- 1 ubuntu ubuntu 468 Oct 14 15:45 myname
```

I left feedback for a couple of things in-line.

Also, regarding your comment:

> Regarding the plugins, I really appreciate your analysis and we should definitively look into those. However, I think the current PR cannot and should not try to solve all problems at once. Usually the plugins will call external function which usually do not make any attempt to secure credentials (database drivers, API implementations, etc). So let's push this down to the plugin level for now and allow for plugins which handle stuff securely, which is already a leap to the current state.

That's a fair point. I do think that these plugins are going to provide a template for future plugins use of secrets, so if pushing down to the plugin level, then lets make them as good as we can and:
* write all zeros to the variable after use of the secret in the plugin. That should be straightforward for the `http` plugin (eg, in `plugins/inputs/http/http.go:gatherURL()` but also verify anywhere else in the plugin). For the `sql` plugin, we can only do this in `Stop()` I think since with the current architecture, it needs to hold onto it
* add a TODO comment in a strategic location for addressing my previous comment: "_The architecture for Start() input plugins that need to start, gather all the inputs, then Stop() all the input plugins that need to stop is a bit problematic and is at odds with how we'd like to use memguard._" I'm not sure what you want to say, but the idea is just mention something about how the architecture needs some tweaking to better support secret-store, perhaps adding a URL to comments in this PR
* add a comment in the `sql` plugin documentation about this limitation of its usage of secret storage (holding onto the unencrypted secret in memory during runtime)

srebhan commented on config/secret.go:
```diff
@@ -0,0 +1,197 @@
+// Initialize the secret content
+func (s *Secret) init(b []byte) {
+	secret := b
+	if unquoted, err := strconv.Unquote(string(b)); err == nil {
```
Fixed in the next push...

srebhan commented on docs/COMMANDS_AND_FLAGS.md:
```diff
@@ -66,3 +67,61 @@ telegraf [flags]
+secret-store with ID `someid`. To also reveal the *values* of all
+the secrets use
+
+`telegraf list-secrets --reveal-secret someid`
```
The idea is to be able to get *all* secrets with values. Using `get-secret` you would need to iterate over the secret-keys... So it is a convenience thing... :-)

srebhan replied:
Can we keep this?

jdstrand replied:
This was an observation, not a blocker. If this is the design you want, feel free to keep it.

srebhan commented on plugins/inputs/sql/sql.go:
```diff
@@ -354,8 +358,12 @@ func (s *SQL) Start(_ telegraf.Accumulator) error {
+	if err != nil {
+		return fmt.Errorf("getting DSN failed: %v", err)
+	}
+	s.Log.Debugf("Connecting to %q...", dsn)
```
Removed the DSN output. Secrets in `http` is not leaked.

srebhan commented:
Thanks @jdstrand for your review! I addressed your comments and would like to keep the `list-secrets` command for convenience... 

> [...] I did a cursory look over the whole PR and nothing jumped out, but am relying on the telegraf team to review those parts.

Yes, that's the plan. After resolving all security issues between us ;-) I'll ask the Telegraf team to review the PR.

> [...] I do think that these plugins are going to provide a template for future plugins use of secrets, so if pushing down to the plugin level, then lets make them as good as we can and:
>
> write all zeros to the variable after use of the secret in the plugin. That should be straightforward for the http plugin (eg, in plugins/inputs/http/http.go:gatherURL() but also verify anywhere else in the plugin). For the sql plugin, we can only do this in Stop() I think since with the current architecture, it needs to hold onto it

I could not find a reliable way to overwrite the `string` with zeros as in Golang `strings` are immutable. Do you know of a way? The only thing I could think of is to return the secret as `[]byte` which can be overwritten and then pass the secret on using
`someFunction(...,string(secret),...)`. Do you think this is a better way? If so, I can modify the code accordingly.

> add a TODO comment in a strategic location for addressing my previous comment: "The architecture for Start() input plugins that need to start, gather all the inputs, then Stop() all the input plugins that need to stop is a bit problematic and is at odds with how we'd like to use memguard." I'm not sure what you want to say, but the idea is just mention something about how the architecture needs some tweaking to better support secret-store, perhaps adding a URL to comments in this PR
add a comment in the sql plugin documentation about this limitation of its usage of secret storage (holding onto the unencrypted secret in memory during runtime)

I don't think this is an issue. As even for the `sql` plugin, we could "delete" the plain-text secret after passing it on to the respective library function. I guess that works for most, if not all, plugins, which brings us back to the question above on how to best "delete/erase" the plain-text stuff... I appreciate any idea!

srebhan commented:
@jdstrand the last commit converts the returned secrets as `[]byte` and erases them ASAP in the plugin functions. I also lock the memory (where available) to prevent swapping of the memory section. Do you like it?

jdstrand reviewed (changes requested):
>> [...] I do think that these plugins are going to provide a template for future plugins use of secrets, so if pushing down to the plugin level, then lets make them as good as we can and:
>> * write all zeros to the variable after use of the secret in the plugin. That should be straightforward for the http plugin (eg, in plugins/inputs/http/http.go:gatherURL() but also verify anywhere else in the plugin). For the sql plugin, we can only do this in Stop() I think since with the current architecture, it needs to hold onto it

> I could not find a reliable way to overwrite the string with zeros as in Golang strings are immutable. Do you know of a way? The only thing I could think of is to return the secret as []byte which can be overwritten and then pass the secret on using
someFunction(...,string(secret),...). Do you think this is a better way? If so, I can modify the code accordingly.

I may not have been clear. `config/secret.go:Get()` is defined as `func (s *Secret) Get() ([]byte, error)` and, for example, `plugins/inputs/http/http.go` does:
```
func (h *HTTP) setRequestAuth(request *http.Request) error {
        username, err := h.Username.Get()
        if err != nil {
                return fmt.Errorf("getting username failed: %v", err)
        }
        defer config.ReleaseSecret(username)
        password, err := h.Password.Get()
        if err != nil {
                return fmt.Errorf("getting password failed: %v", err)
        }
        defer config.ReleaseSecret(password)
        if len(username) != 0 || len(password) != 0 {
                request.SetBasicAuth(string(username), string(password))                                 
        }
        return nil
}
```

I'm only suggesting that you do something like:
```
...
                request.SetBasicAuth(string(username), string(password))
                // These local slices are backed by dynamically allocated arrays, so zero them out
                // manually so they don't persist in GC. Note: slices can also be on the heap based on
                // what the Go compiler decides and if what was append()ed to them exceeded the
                // original allocation. This also makes the window smaller for dumpable and swapped
                // memory.
                for i, _ := range username {
                        username[i] = '0'
                }
                for i, _ := range password {
                        password[i] = '0'
                }
```

Feel free to make this a helper function since the slice has an internal pointer to the backing array and this would access the backing array directly.

Hopefully this helps!

srebhan commented on config/util.go:
```diff
@@ -0,0 +1,23 @@
+		if bytes.HasPrefix(b, []byte(`"`)) && bytes.HasSuffix(b, []byte(`"`)) {
+			return b[1 : len(b)-1]
+		}
+	}
```
Thanks for your review @jdstrand! Here are my answers to your points: ;-)

> `"""'''foo'''"""` returns `'''foo'''` whereas `'''"""foo"""'''` returns foo. Perhaps don't worry about `"""'''` or `'''"""` and use an `else if`?
> similar for "'foo'" and '"foo"'. Maybe use an else if?

The current behavior is that `"""'''foo'''"""` returns `'''foo'''` and `'''"""foo"""'''` returns `"""foo"""`, which is what I would expect. Anyway, I added those examples to the [unit-test](https://github.com/influxdata/telegraf/blob/9f7730ea6bd3cbd3d924c1f8c3e1a654132681c7/config/secret_test.go#L170). Likewise for mixed single quotes.
We don't need `else if`s in the `unquoteTomlString()` as we directly `return` on the first matching condition, so everything else is only executed if nothing matched previously...

> """"foo"""" returns foo. Perhaps the outer conditional should be else if too?

I tested and both quadruple single-quotes as well as double-quotes will trigger a TOML parsing error. Do you want me to add a test for this?

> It would be good to have some table driven tests for `unquoteTomlString()`
That's what `TestSecretUnquote()` in `secret_test.go` does in combination with the TOML parser, which will catch some of the odd cases. Do you want me to add another test for the underlying `unquoteTomlString()` function?

jdstrand replied:
Oh! I missed that each of those 'if's had return statements. Whoops, hehehe :) `secret_test.go` is fine. Thanks for adding those.

jdstrand reviewed (approved):
> @jdstrand regarding zeroing the secret's byte array, this is what the `defer config.ReleaseSecret(...)` lines do. They read
> 
> ```go
> func ReleaseSecret(secret []byte) error {
> 	memguard.WipeBytes(secret)
> 	return nil
> }
> ```
> 
> and on Linux the byte-array is `mlock`ed between `Get()` and `ReleaseSecret()` to prevent swapping it out. Sorry, but I do not get what you want me to do beyond that!? :-(

No, you are right. Part of the problem for me was that there was a bit of a wait between reviews and I lost some context that I thought I had (sorry). The name `ReleaseSecret` for some reason didn't remind me that it is also wiping (but that's not your fault; could be named `ReleaseAndWipeSecret()` but this is not a blocker). But hey, that is part of what reviews are for: getting people on the same page! :) 

Looking at `ReleaseSecret()`, it is already doing what I suggested when it calls `Wipe()` (and like you said, on Linux, between `protect()` and `ReleaseSecret()` it is also managing `mlock`) so there is no more code to write in terms of zeroing out the secret.

What I recalled thinking through and I mistakenly held up `http.go` as a place to do this in the last review, is that, depending on the plugin, it might be useful to not wait for the `defer` statement to do this in an effort to make the window even smaller. In `http.go`, it is fine to do in a `defer` since we're at the end of the function when done with username and password anyway. For `sql.go`, you should be able to call `ReleaseSecret()` immediately after the `Open()` instead of using a `defer` and waiting until all the statements are run. This should be possible since the connection is already open when running the statements. This is not a blocker; it would be nice to do that, but please be sure to test.

I'm going to Approve as is. Thanks for bearing with me on the review!

srebhan commented:
I'll check with the Telegraf team to find a better name for `ReleaseSecret()`... Maybe `WipeSecret()` or similar. Regarding the use of `defer`, I always like to use `defer` because it saves you in case of branching. Like in the `sql` example, the `Open()` statement could fail and we might return without calling `ReleaseSecret()` which causes the memory-page to stay locked. But in general I agree that shorter is better. Also tried to do this other plugins (see #12067)... If you don't mind giving your approval I can trigger the team to start their reviews...

jdstrand commented:
> I'll check with the Telegraf team to find a better name for `ReleaseSecret()`... Maybe `WipeSecret()` or similar. Regarding the use of `defer`, I always like to use `defer` because it saves you in case of branching. Like in the `sql` example, the `Open()` statement could fail and we might return without calling `ReleaseSecret()` which causes the memory-page to stay locked. But in general I agree that shorter is better. Also tried to do this other plugins (see #12067)... If you don't mind giving your approval I can trigger the team to start their reviews...

Yes, this is what makes `defer` great and I totally agree in the normal case. In this one case, it might be worth breaking that convention (with a justifying code comment) and calling `ReleaseSecret()` in the `if` conditional as well as immediately outside. Like I said though, this is not a blocker and I approved as is.

powersj commented on cmd/telegraf/telegraf.go:
```diff
@@ -221,16 +245,18 @@ func (t *Telegraf) runAgent(ctx context.Context) error {
 
 	log.Printf("I! Starting Telegraf %s%s", internal.Version, internal.Customized)
-	log.Printf("I! Available plugins: %d inputs, %d aggregators, %d processors, %d parsers, %d outputs",
+	log.Printf("I! Available plugins: %d inputs, %d aggregators, %d processors, %d parsers, %d outputs %d, secret-stores",
```
```suggestion
	log.Printf("I! Available plugins: %d inputs, %d aggregators, %d processors, %d parsers, %d outputs, %d secret-stores",
```

powersj reviewed (commented):
This is awesome! While a large PR in itself, it is really more straightforward than I thought it would be. I do want to try this on macOS, but I wanted to put this in now so you can see it earlier rather than later.

I have a couple inline changes/questions, but all of them minor. I think the great reviews got this setup for success. 

My biggest change is I do want to see us using the [noun/verb layout](https://clig.dev/#subcommands) of subcommands versus adding lots of various subcommands that should be grouped. This means:

```
telegraf secret [get|set|list]
telegraf secretstore list
```

As an example:

```
Before:
telegraf get-secret someid a_secret_key

After:
telegraf secret get someid a_secret_key
```

Let me know what you think!

powersj reviewed (commented):
Two comments, should be quick and we land

srebhan commented on docs/COMMANDS_AND_FLAGS.md:
````diff
@@ -55,3 +61,62 @@ If a user only wanted certain inputs or outputs, then the filters can be used:
+secret-store with ID `someid`. To also reveal the *values* of all
+the secrets use
+
+`telegraf secrets list --reveal-secret someid`
+
+You can also pass a list of secret-stores and the command will
+print all secrets in those stores.
+
+To access the *value* of a specific secret you can use
+
+`telegraf secrets get someid a_secret_key`
+
+to output the *value* of the secret `a_secret_key` in the secret-store
+with ID `someid`.
+
+All commands above will *read* secrets stored in the given store(s). To add
+or modify keys use the `set` command
+
+`telegraf secrets set someid a_secret_key the_new_value`
+
+to add or overwrite a secret named `a_secret_key` with the value
+`the_new_value` in the secret-store with ID `someid`. If the secret with
+the given key did not exist, it will be created.
+
+You may now reference the secret in your telegraf configuration anywhere
+you would normally use a password or secret. If your configuration has e.g.
+
+```toml
+  token = "<your secret token>"
+```
+
+ ou can replace it with a reference in the form
+`@{<secret store id>:<secret name>}`. For the example above you would write
+
+```toml
+  token = @{someid:a_secret_key}
+```
+
+to reference the secret with the key `a_secret_key` of the secret-store with
+ID `someid`.
````
So you mean remove this whole section?

powersj replied:
Essentially yes - I don't want us to have to maintain a list of CLI options more than the subcommands. And instead push users to using the CLI itself to get the help they need

powersj reviewed (approved):
Looks great!

Hipska reviewed (commented):
A bummer I missed this PR while it was still open, but here some remarks anyway. I hope I can still get a reply on them or they are addressed in a future PR.

First of all, there isn't a `docs/SECRETSTORES.md` to explain what is required when creating a new plugin just like there is for any other type of plugin. Beware to also reference this in `README.md#documentation`.

srebhan commented on plugins/secretstores/jose/jose.go:
```diff
@@ -0,0 +1,108 @@
+		return errors.New("id missing")
+	}
+
+	passwd, err := j.Password.Get()
+	if err != nil {
+		return fmt.Errorf("getting password failed: %v", err)
+	}
```
It is not absolutely necessary as the above (first line in `Init()`) destruction of the secret will also wipe that buffer. But added it here for completeness.

srebhan commented on plugins/secretstores/jose/sample.conf:
```diff
@@ -0,0 +1,13 @@
+  id = "secretstore"
+
+  ## Directory for storing the secrets
+  # path = "secrets"
```
Made path a mandatory option.

srebhan commented on plugins/inputs/http/http.go:
```diff
@@ -184,6 +185,23 @@ func (h *HTTP) gatherURL(
+	if err != nil {
+		return fmt.Errorf("getting username failed: %v", err)
+	}
+	defer config.ReleaseSecret(username)
```
The (clear-text) secret given to you is a byte-array. To protect this clear-text to be e.g. swapped-out to disk, we protect the memory region and want to release this protection. Furthermore, we want to safely wipe the secret from memory once it is not used anymore. Therefore, we need some indication when to do this (i.e. the release). We could of course also clean this up on garbage-collection, but this would make the plain-text secret to stay in memory for an unforeseeable time...
